src/seismometer/controls/decorators.py

```diff
@@ -2,13 +2,79 @@
 """
 As for all of the utils subpackage, no other custom subpackages should be referenced.
 """
-import builtins
+import hashlib
+import logging
+import os
+import shutil
 from functools import wraps
+from inspect import signature
+from pathlib import Path
 from typing import Any, Callable
 
-from IPython.display import display
+from IPython.display import HTML
 from ipywidgets import Widget
 
+logger = logging.getLogger("seismometer")
+SEISMOMETER_CACHE_DIR = Path("./.seismometer_cache")
+SEISMOMETER_CACHE_ENABLED = os.getenv("SEISMOMETER_CACHE_ENABLED", "") != ""
+
+
+def disk_cached_html_segment(func: Callable[..., HTML]) -> HTML:
+    """
+    Caching decorator for HTML data
+
+    Parameters
+    ----------
+    func : Callable[..., HTML]
+        function that generates an HTML component
+
+    Returns
+    -------
+    HTML
+        displayable HTML component
+
+    Raises
+    ------
+    TypeError
+        If called with a function that does not annotate its return type as HTML, or that returns a non-HTML object.
+    """
+    sig = signature(func)
+    if sig.return_annotation != HTML:
+        raise TypeError("The function must return {HTML} object.")
+
+    if not SEISMOMETER_CACHE_ENABLED:
+        return func
+
+    @wraps(func)
+    def wrapped_func(*args, **kwargs):
+        arg_str = str((args, frozenset(kwargs.items())))
```

Comment:
We already chatted but to get it documented, dataframes are annoying.  
the string representation hides rows/columns which can cause collisions + reading parquet doesnt seem to guarantee column order

Reply:
updated to have a safe hash based on the dataframe contents.